### PR TITLE
 Accept Brotli-compressed HTTP responses #8156

### PR DESCRIPTION
**Pull request by nxnfufunezn**
r? @jdm

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/8192)

### CODE REVIEW

**Comment by jdm** [comment]
This looks like what I expected; thanks! Have you tested it manually? Easiest way to verify it's working is to comment out the branch on 427 and find a page that serves Brotli-encoded data (and therefore is full of garbage), then uncomment that branch and see that the page shows up correctly.

**Comment by nxnfufunezn** [comment]
@jdm Is there any site  that serves Brotli-encoded data ? Or how can I verify this any other way ?

**Comment by nxnfufunezn** [comment]
@jdm It should be ```Accept-Encoding: br``` instead of ```Accept-Encoding: Brotli``` am I right ?

File: components/net/Cargo.toml [hunk 1]
```diff
@@ -16,6 +16,9 @@ path = "../util"
 [dependencies.devtools_traits]
 path = "../devtools_traits"
 
+[dependencies.brotli]
+git = "https://github.com/ende76/brotli-rs"
```
Comment (frewsxcv) [inline components/net/Cargo.toml]: Don't forget to add the Cargo.lock files

**Comment by jdm** [comment]
@nxnfufunezn You're right! Aso, https://http2.cloudflare.com/ supports Brotli encoding according to my Firefox network inspector.

**Comment by nxnfufunezn** [comment]
Oh I am using this python script : https://gist.github.com/nxnfufunezn/d3fe9c805004a94a262d for testing...

**Comment by nxnfufunezn** [comment]
Also, I think I need to add brotli here also  https://github.com/servo/servo/blob/master/components/net/http_loader.rs#L156

**Comment by jdm** [comment]
Even better! That looks like something we could turn into an automated test, like this one: http://mxr.mozilla.org/servo/source/tests/wpt/web-platform-tests/XMLHttpRequest/resources/gzip.py?force=1

**Comment by jdm** [comment]
Yep, you're correct.

**Comment by nxnfufunezn** [comment]
okay I will do that also.. :+1:

**Comment by nxnfufunezn** [comment]
@jdm It works now..

**Comment by nxnfufunezn** [comment]
Here is what I have done so far https://gist.github.com/nxnfufunezn/220bd5dc1dfb590bb8d6

**Comment by jdm** [comment]
@nxnfufunezn That looks really good!

**Comment by nxnfufunezn** [comment]
@jdm

**Comment by jdm** [comment]
Looks great! We just need to rebase this to get the proper manifest output, as described in the comment. Please also include the Cargo.lock changes, like @frewsxcv pointed out.
-S-awaiting-review +S-needs-rebase 


---

Reviewed 3 of 3 files at r1, 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

<sup>**[tests/wpt/mozilla/meta/MANIFEST.json, line 3806 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8192#-K1nPbJtFAfneLV2UnSj)** ([raw file](https://github.com/servo/servo/blob/285e29c06637f31a8b8a27c2e454468717924ebd/tests/wpt/mozilla/meta/MANIFEST.json#L3806)):</sup>
If you rebase to master and run `./mach test-wpt --update-manifest` again, this should produce difference results since #8185 merged.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8192)

**Comment by frewsxcv** [comment]
<sup>**[tests/wpt/mozilla/tests/mozilla/resources/brotli.py, line 5 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8192#-K1nX2lPEvaF8yKOOe5u)** ([raw file](https://github.com/servo/servo/blob/58207c0701c847e1cc6e5ffb63a111a4035af128/tests/wpt/mozilla/tests/mozilla/resources/brotli.py#L5)):</sup>
This can be written:

```
output = request.GET.get("content", request.body)
```

https://docs.python.org/2/library/stdtypes.html#dict.get


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8192)

**Comment by nxnfufunezn** [comment]
<sup>**[tests/wpt/mozilla/tests/mozilla/resources/brotli.py, line 5 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8192#-K1nX2lPEvaF8yKOOe5u)** ([raw file](https://github.com/servo/servo/blob/58207c0701c847e1cc6e5ffb63a111a4035af128/tests/wpt/mozilla/tests/mozilla/resources/brotli.py#L5)):</sup>
So I don't need brotli compressed data there ?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8192)

**Comment by frewsxcv** [comment]
<sup>**[tests/wpt/mozilla/tests/mozilla/resources/brotli.py, line 5 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8192#-K1nX2lPEvaF8yKOOe5u)** ([raw file](https://github.com/servo/servo/blob/58207c0701c847e1cc6e5ffb63a111a4035af128/tests/wpt/mozilla/tests/mozilla/resources/brotli.py#L5)):</sup>
What do you mean by that? My code block is identical in functionality to your `if..else` statement


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8192)

**Comment by nxnfufunezn** [comment]
<sup>**[tests/wpt/mozilla/tests/mozilla/resources/brotli.py, line 5 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8192#-K1nX2lPEvaF8yKOOe5u)** ([raw file](https://github.com/servo/servo/blob/58207c0701c847e1cc6e5ffb63a111a4035af128/tests/wpt/mozilla/tests/mozilla/resources/brotli.py#L5)):</sup>
oh.. I got it..


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8192)

**Comment by nxnfufunezn** [comment]
<sup>**[tests/wpt/mozilla/tests/mozilla/resources/brotli.py, line 5 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8192#-K1nX2lPEvaF8yKOOe5u)** ([raw file](https://github.com/servo/servo/blob/58207c0701c847e1cc6e5ffb63a111a4035af128/tests/wpt/mozilla/tests/mozilla/resources/brotli.py#L5)):</sup>
Done.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8192)

**Comment by nxnfufunezn** [comment]
ping @jdm @frewsxcv

**Comment by jdm** [comment]
@nxnfufunezn What commands did you run that caused so many changes to the Cargo.lock files?

**Comment by nxnfufunezn** [comment]
./mach cargo-update

**Comment by nxnfufunezn** [comment]
reverted those cargo.lock changes ! @jdm

**Comment by nxnfufunezn** [comment]
@jdm Any changes that I have to make?

**Comment by frewsxcv** [comment]
All of those changes in `tests/wpt/mozilla/meta/MANIFEST.json` seem concerning

**Comment by nxnfufunezn** [comment]
All I did was ./mach create-wpt /path/to/response-brotli.htm to generate manifest file.

**Comment by nxnfufunezn** [comment]
@frewsxcv All I did was ./mach create-wpt /path/to/response-brotli.htm to generate manifest file.

**Comment by nxnfufunezn** [comment]
Again tried generating manifest got same thing...

**Comment by Ms2ger** [comment]
Yeah, not your fault. Could you revert the changes manually, though?

**Comment by nxnfufunezn** [comment]
all the manifest changes right ?

**Comment by nxnfufunezn** [comment]
@Ms2ger done!

**Comment by frewsxcv** [comment]
@nxnfufunezn I'm going to fix the broken entries in the manifest.json before moving forward on this, sorry about the wait!

**Comment by frewsxcv** [comment]
https://github.com/servo/servo/pull/8268

**Comment by frewsxcv** [comment]
@nxnfufunezn Run the `create-wpt` command again, but don't commit the XML mentions (as mentioned here: https://github.com/servo/servo/issues/8270)

**Comment by nxnfufunezn** [comment]
@frewsxcv Done!

**Comment by frewsxcv** [comment]
It looks like this breaks a few unit tests?

http://build.servo.org/builders/mac-dev-ref-unit/builds/131/steps/shell_1/logs/stdio

**Comment by nxnfufunezn** [comment]
I think I need to update unit-test tests/unit/net/http_loader.rs accordingly

**Comment by nxnfufunezn** [comment]
Brotli needs to be added here https://github.com/servo/servo/blob/master/tests/unit/net/http_loader.rs#L310 right ?

**Comment by jdm** [comment]
Sorry for going silent; I was at a conference for two days.
-S-awaiting-review +S-needs-code-changes 


---

Reviewed 7 of 7 files at r3, 6 of 6 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

<sup>**[components/net/http_loader.rs, line 427 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8192#-K1vaEy83RezSzcQpZB4)** ([raw file](https://github.com/servo/servo/blob/d1a0431e7da4423fd207cbabc0e4446b0c778eb4/components/net/http_loader.rs#L427)):</sup>
We should do `Some(Encoding::EncodingExt(ref ext)) if ext == "br" => {`, otherwise this will capture any extension value.


---

<sup>**[components/net/http_loader.rs, line 161 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8192#-K1va8kk_ChIvxvqqU3M)** ([raw file](https://github.com/servo/servo/blob/58207c0701c847e1cc6e5ffb63a111a4035af128/components/net/http_loader.rs#L161)):</sup>
nit: `} else {`


---

<sup>**[tests/wpt/mozilla/tests/mozilla/resources/brotli.py, line 2 \[r3\]](https://reviewable.io:443/reviews/servo/servo/8192#-K1v_9ybn6BHyS3GWiql)** ([raw file](https://github.com/servo/servo/blob/bd3ff7d0ca216a4b8b3347004151ae38c14c5fe0/tests/wpt/mozilla/tests/mozilla/resources/brotli.py#L2)):</sup>
This line should be removed, since it's not doing anything (we overwrite output after this).


---

<sup>**[ports/cef/Cargo.lock, line 131 \[r3\]](https://reviewable.io:443/reviews/servo/servo/8192#-K1v_gMZfoHt_aJ6JGDC)** ([raw file](https://github.com/servo/servo/blob/bd3ff7d0ca216a4b8b3347004151ae38c14c5fe0/ports/cef/Cargo.lock#L131)):</sup>
Please run `./mach cargo-update -p brotli --precise 0.3.17` to ensure that the CEF and Gonk Cargo.lock files are updated appropriately.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8192)

**Comment by nxnfufunezn** [comment]
<sup>**[ports/cef/Cargo.lock, line 131 \[r3\]](https://reviewable.io:443/reviews/servo/servo/8192#-K1v_gMZfoHt_aJ6JGDC)** ([raw file](https://github.com/servo/servo/blob/bd3ff7d0ca216a4b8b3347004151ae38c14c5fe0/ports/cef/Cargo.lock#L131)):</sup>
do I need to add brotli dependency in CEF and Gonk Cargo.toml file before this ?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8192)

**Comment by nxnfufunezn** [comment]
<sup>**[ports/cef/Cargo.lock, line 131 \[r3\]](https://reviewable.io:443/reviews/servo/servo/8192#-K1v_gMZfoHt_aJ6JGDC)** ([raw file](https://github.com/servo/servo/blob/bd3ff7d0ca216a4b8b3347004151ae38c14c5fe0/ports/cef/Cargo.lock#L131)):</sup>
because I am getting this --

$ ./mach cargo-update -p brotli --precise 0.3.17

components/servo
thread '<main>' panicked at 'index 0 and/or 8 in `0.3.17` do not lie on character boundary', ../src/libcore/str/mod.rs:1284
ports/cef
package id specification `brotli` matched no packages
ports/gonk
package id specification `brotli` matched no packages


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8192)

**Comment by nxnfufunezn** [comment]
I guess in case of git repos we have to use SHA instead of version number ..

**Comment by jdm** [comment]
Oh, yes, try using a SHA instead. I filed https://github.com/rust-lang/cargo/issues/2096 for the poor error message.

**Comment by nxnfufunezn** [comment]
@jdm applied all the changes..

**Comment by eefriedman** [comment]
When you're updating cargo dependencies, use `./mach update-cargo`; there are actually three different Cargo.lock files in the Servo tree, and that calls `cargo update1`all of them.

**Comment by jdm** [comment]
Just waiting on the changes to the other lock files.
-S-awaiting-review +S-needs-code-changes 


---

Reviewed 6 of 6 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8192)

**Comment by nxnfufunezn** [comment]
@eefriedman ah so it was ```fish ./mach update-cargo ``` ... I was doing ```fish ./mach cargo-update ```

**Comment by jdm** [comment]
Odd. Those are interchangeable: http://mxr.mozilla.org/servo/source/python/servo/devenv_commands.py#44

**Comment by frewsxcv** [comment]
@bors-servo retry

#7787

**Comment by jdm** [comment]
@bors-servo: retry
* #7731

**Comment by eefriedman** [comment]
@bors-servo retry

#7787

**Pull request merged**